src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/WebSocketsTransport.cs

```diff
@@ -68,41 +69,54 @@ public WebSocketsTransport(WebSocketOptions options, IDuplexPipe application, De
             var receiving = StartReceiving(socket);
             var sending = StartSending(socket);
 
-            // Wait for something to shut down.
-            var trigger = await Task.WhenAny(
-                receiving,
-                sending);
+            // Wait for send or receive to complete
+            var trigger = await Task.WhenAny(receiving, sending);
 
-            var failed = trigger.IsCanceled || trigger.IsFaulted;
-            var task = Task.CompletedTask;
             if (trigger == receiving)
             {
-                task = sending;
                 _logger.WaitingForSend();
+
+                // We're waiting for the application to finish and there are 2 things it could be doing
+                // 1. Waiting for application data
+                // 2. Waiting for a websocket send to complete
+
+                // Cancel the application so that ReadAsync yields
+                _application.Input.CancelPendingRead();
+
+                var resultTask = await Task.WhenAny(sending, Task.Delay(_options.CloseTimeout));
+
+                if (resultTask != sending)
+                {
+                    // We timed out so now we're in ungraceful shutdown mode
+                    _logger.CloseTimedOut();
+
+                    // Abort the websocket if we're stuck in a pending send to the client
+                    _aborted = true;
+
+                    socket.Abort();
+                }
             }
             else
             {
-                task = receiving;
                 _logger.WaitingForClose();
-            }
 
-            await socket.CloseOutputAsync(failed ? WebSocketCloseStatus.InternalServerError : WebSocketCloseStatus.NormalClosure, "", CancellationToken.None);
+                // We're waiting on the websocket to close and there are 2 things it could be doing
+                // 1. Waiting for websocket data
+                // 2. Waiting on a flush to complete (backpressure being applied)
 
-            var resultTask = await Task.WhenAny(task, Task.Delay(_options.CloseTimeout));
+                var resultTask = await Task.WhenAny(receiving, Task.Delay(_options.CloseTimeout));
```

Comment:
Given our history with `Task.Delay`, should we put a cancellation token in to cancel the `Delay` if it isn't needed (which is the common case)?

Reply:
Allocations everywhere! Sure, I'll make this change

Reply:
Well I hear all your changes have been regressing perf anyway :trollface: .

If you think the allocations are more of a problem than the timers in the queue I'm certainly fine with not using the token.